### ISSUE TITLE
Unable to know which field failed to parse if the response for a field defined with non-null is null

### ISSUE THREAD

**Issue by kubode** [apollographql/apollo-kotlin#3912]
**Version**
3.1.0

**Use case**

We are about to make a destructive schema change that will make fields that were non-null to nullable.
This change will cause the released app to throw an exception if the response is actually null.
This is the expected behavior, but we cannot determine which field failed to parse from the stack trace.

Stacktrace

```
com.apollographql.apollo3.exception.ApolloParseException: Failed to parse GraphQL http network response
    at com.apollographql.apollo3.network.http.HttpNetworkTransport$Companion.wrapThrowableIfNeeded(HttpNetworkTransport.kt:193)
    at com.apollographql.apollo3.network.http.HttpNetworkTransport$Companion.access$wrapThrowableIfNeeded(HttpNetworkTransport.kt:187)
    at com.apollographql.apollo3.network.http.HttpNetworkTransport$execute$1$response$1.invoke(HttpNetworkTransport.kt:79)
    at com.apollographql.apollo3.network.http.HttpNetworkTransport$execute$1$response$1.invoke(HttpNetworkTransport.kt:72)
    at com.apollographql.apollo3.internal.NonMainWorker.doWork(NonMainWorker.kt:5)
    at com.apollographql.apollo3.network.http.HttpNetworkTransport$execute$1.invokeSuspend(HttpNetworkTransport.kt:72)
    at com.apollographql.apollo3.network.http.HttpNetworkTransport$execute$1.invoke(Unknown Source:8)
    at com.apollographql.apollo3.network.http.HttpNetworkTransport$execute$1.invoke(Unknown Source:4)
    at kotlinx.coroutines.flow.SafeFlow.collectSafely(Builders.kt:61)
    at kotlinx.coroutines.flow.AbstractFlow.collect(Flow.kt:227)
    at kotlinx.coroutines.flow.internal.ChannelFlowOperatorImpl.flowCollect(ChannelFlow.kt:195)
    at kotlinx.coroutines.flow.internal.ChannelFlowOperator.collect$suspendImpl(ChannelFlow.kt:167)
    at kotlinx.coroutines.flow.internal.ChannelFlowOperator.collect(Unknown Source:0)
    at kotlinx.coroutines.flow.FlowKt__TransformKt$onEach$$inlined$unsafeTransform$1.collect(SafeCollector.common.kt:113)
    at kotlinx.coroutines.flow.FlowKt__ReduceKt.single(Reduce.kt:57)
    at kotlinx.coroutines.flow.FlowKt.single(Unknown Source:1)
    at com.apollographql.apollo3.cache.normalized.internal.ApolloCacheInterceptor.readFromNetwork(ApolloCacheInterceptor.kt:211)
    at com.apollographql.apollo3.cache.normalized.internal.ApolloCacheInterceptor.access$readFromNetwork(ApolloCacheInterceptor.kt:31)
    at com.apollographql.apollo3.cache.normalized.internal.ApolloCacheInterceptor$interceptQuery$1.invokeSuspend(ApolloCacheInterceptor.kt:168)
    at com.apollographql.apollo3.cache.normalized.internal.ApolloCacheInterceptor$interceptQuery$1.invoke(Unknown Source:8)
    at com.apollographql.apollo3.cache.normalized.internal.ApolloCacheInterceptor$interceptQuery$1.invoke(Unknown Source:4)
    at kotlinx.coroutines.flow.SafeFlow.collectSafely(Builders.kt:61)
    at kotlinx.coroutines.flow.AbstractFlow.collect(Flow.kt:227)
    at kotlinx.coroutines.flow.internal.ChannelFlowOperatorImpl.flowCollect(ChannelFlow.kt:195)
    at kotlinx.coroutines.flow.internal.ChannelFlowOperator.collectTo$suspendImpl(ChannelFlow.kt:157)
    at kotlinx.coroutines.flow.internal.ChannelFlowOperator.collectTo(Unknown Source:0)
    at kotlinx.coroutines.flow.internal.ChannelFlow$collectToFun$1.invokeSuspend(ChannelFlow.kt:60)
    at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
    at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:106)
    at kotlinx.coroutines.internal.LimitedDispatcher.run(LimitedDispatcher.kt:39)
    at kotlinx.coroutines.scheduling.TaskImpl.run(Tasks.kt:95)
    at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(CoroutineScheduler.kt:571)
    at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.executeTask(CoroutineScheduler.kt:750)
    at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.runWorker(CoroutineScheduler.kt:678)
    at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.kt:665)
Caused by: java.lang.IllegalStateException: Check failed.
    at com.apollographql.apollo3.api.json.MapJsonReader.beginObject(MapJsonReader.kt:108)
    at com.apollographql.apollo3.api.json.MapJsonReader.beginObject(MapJsonReader.kt:29)
    at com.apollographql.apollo3.api.ObjectAdapter.fromJson(Adapters.kt:229)
...
    at com.apollographql.apollo3.api.ObjectAdapter.fromJson(Adapters.kt:230)
    at com.apollographql.apollo3.api.NullableAdapter.fromJson(Adapters.kt:58)
    at com.apollographql.apollo3.api.internal.ResponseParser.parse(ResponseParser.kt:35)
    at com.apollographql.apollo3.api.Operations.parseJsonResponse(Operations.kt:59)
    at com.apollographql.apollo3.network.http.HttpNetworkTransport$execute$1$response$1.invoke(HttpNetworkTransport.kt:74)
	... 32 more
```

**Describe the solution you'd like**

Can you improve the exception message when parsing fails?

Such as `IllegalStateException: Failed to parse ${path}. Expected an object but was null.`

**Comment by martinbonnin** [comment]
Indeed, [`MapJsonReader`](https://github.com/apollographql/apollo-kotlin/blob/932f6fd8c5818274f87a395d059deeb8e3e6142b/apollo-api/src/commonMain/kotlin/com/apollographql/apollo3/api/json/MapJsonReader.kt#L29) doesn't have nice error reporting like [`BufferedSourceJsonReader`](https://github.com/apollographql/apollo-kotlin/blob/31c873c5abbc4873c98e31ba17da7c16e90109ef/apollo-api/src/commonMain/kotlin/com/apollographql/apollo3/api/json/BufferedSourceJsonReader.kt#L106). We'd need to add a `MapJsonReader.path()` and better error messages in the different `nextFoo()`, `beginObject()` and `beginArray()`. Do you want to open a PR?

**Comment by kubode** [comment]
I'll try this week.
Should I refer to the `BufferedSourceJsonReader` for implementation?

**Comment by martinbonnin** [comment]
> Should I refer to the BufferedSourceJsonReader for implementation?

Mostly yes although the state/path doesn't really use the same data structures. But the ideas are similar.

**Comment by martinbonnin** [comment]
It should be all good now with https://github.com/apollographql/apollo-kotlin/commit/897b8a386615505efec0aaaf761e98fc3036d05c and following commits. 

We should still add a couple of unit tests before closing this issue.

**Comment by kubode** [comment]
I'm sorry I haven't been able to work on it, even though I said I would try 🙇

**Comment by BoD** [comment]
No worry at all! We actually needed `MapJsonAdapter.getPath` for a different reason!

**Comment by BoD** [comment]
Available in v3.3.0.
